### DISCUSSION TITLE
"EOFError: Ran out of input" when packaging a Python script with PyInstaller

### DISCUSSION THREAD

**Discussion by CarloPoletto** [pyinstaller/pyinstaller#7409] [Help] [answered]
I'm developing an application for Windows operating systems written in Python 3.8 and which makes use of the nnunet library (https://pypi.org/project/nnunet/) which uses multiprocessing. I have tested the script and it works correctly.

Now I'm trying to package everything with pyinstaller v5.7.0. The creation of the .exe is successful but when I run it I get the following error:

```
Traceback (most recent call last):
  File "main.py", line 344, in <module>
  File "nnunet\inference\predict.py", line 694, in predict_from_folder
  File "nnunet\inference\predict.py", line 496, in predict_cases_fastest
  File "nnunet\inference\predict.py", line 123, in preprocess_multithreaded
  File "multiprocess\process.py", line 121, in start
  File "multiprocess\context.py", line 224, in _Popen
  File "multiprocess\context.py", line 327, in _Popen
  File "multiprocess\popen_spawn_win32.py", line 93, in __init__
  File "multiprocess\reduction.py", line 70, in dump
  File "dill\_dill.py", line 394, in dump
  File "pickle.py", line 487, in dump
  File "dill\_dill.py", line 388, in save
  File "pickle.py", line 603, in save
  File "pickle.py", line 717, in save_reduce
  File "dill\_dill.py", line 388, in save
  File "pickle.py", line 560, in save
  File "dill\_dill.py", line 1186, in save_module_dict
  File "pickle.py", line 971, in save_dict
Traceback (most recent call last):
  File "main.py", line 341, in <module>
  File "pickle.py", line 997, in _batch_setitems
  File "D:\MyProject\venv\Lib\site-packages\PyInstaller\hooks\rthooks\pyi_rth_multiprocessing.py", line 49, in _freeze_support
  File "dill\_dill.py", line 388, in save
    spawn.spawn_main(**kwds)
  File "pickle.py", line 560, in save
  File "pickle.py", line 901, in save_tuple
  File "dill\_dill.py", line 388, in save
  File "multiprocessing\spawn.py", line 116, in spawn_main
  File "pickle.py", line 560, in save
  File "multiprocessing\spawn.py", line 126, in _main
  File "dill\_dill.py", line 1427, in save_instancemethod0
EOFError: Ran out of input
[588] Failed to ex  File "pickle.py", line 692, in save_reduce
ecute script 'main' d  File "dill\_dill.py", line 388, in save
ue to unhandled  File "pickle.py", line 560, in save
 exception!
  File "pickle.py", line 886, in save_tuple
  File "dill\_dill.py", line 388, in save
  File "pickle.py", line 603, in save
  File "pickle.py", line 717, in save_reduce
  File "dill\_dill.py", line 388, in save
  File "pickle.py", line 560, in save
  File "dill\_dill.py", line 1186, in save_module_dict
  File "pickle.py", line 971, in save_dict
  File "pickle.py", line 997, in _batch_setitems
  File "dill\_dill.py", line 388, in save
  File "pickle.py", line 603, in save
  File "pickle.py", line 687, in save_reduce
  File "dill\_dill.py", line 388, in save
  File "pickle.py", line 560, in save
  File "dill\_dill.py", line 1698, in save_type
  File "dill\_dill.py", line 1070, in _save_with_postproc
  File "pickle.py", line 692, in save_reduce
  File "dill\_dill.py", line 388, in save
  File "pickle.py", line 560, in save
  File "pickle.py", line 901, in save_tuple
  File "dill\_dill.py", line 388, in save
  File "pickle.py", line 560, in save
  File "pickle.py", line 886, in save_tuple
  File "dill\_dill.py", line 388, in save
  File "pickle.py", line 560, in save
  File "dill\_dill.py", line 1698, in save_type
  File "dill\_dill.py", line 1084, in _save_with_postproc
  File "pickle.py", line 997, in _batch_setitems
  File "dill\_dill.py", line 388, in save
  File "pickle.py", line 603, in save
  File "pickle.py", line 717, in save_reduce
  File "dill\_dill.py", line 388, in save
  File "pickle.py", line 560, in save
  File "dill\_dill.py", line 1186, in save_module_dict
  File "pickle.py", line 971, in save_dict
  File "pickle.py", line 997, in _batch_setitems
  File "dill\_dill.py", line 388, in save
  File "pickle.py", line 603, in save
  File "pickle.py", line 717, in save_reduce
  File "dill\_dill.py", line 388, in save
  File "pickle.py", line 560, in save
  File "dill\_dill.py", line 1186, in save_module_dict
  File "pickle.py", line 971, in save_dict
  File "pickle.py", line 997, in _batch_setitems
  File "dill\_dill.py", line 388, in save
  File "pickle.py", line 578, in save
  File "PyInstaller\loader\pyimod01_archive.py", line 76, in __getattr__
AssertionError
[4392] Failed to execute script 'main' due to unhandled exception!
```

Below is the code of my python script:

```
#==============================
# main.py
#==============================

from multiprocessing import freeze_support
from nnunet.inference.predict import predict_from_folder

if __name__ == "__main__":
    freeze_support()
    ...
    predict_from_folder(...)
    ...
```

Below is the code of the nnunet library that triggers the error:

```
#==============================
# nnunet\inference\predict.py
#==============================

def preprocess_multithreaded(trainer, list_of_lists, output_files, num_processes=2, segs_from_prev_stage=None):
    if segs_from_prev_stage is None:
        segs_from_prev_stage = [None] * len(list_of_lists)

    num_processes = min(len(list_of_lists), num_processes)

    classes = list(range(1, trainer.num_classes))
    assert isinstance(trainer, nnUNetTrainer)
    q = Queue(1)
    processes = []
    for i in range(num_processes):
        pr = Process(
            target=preprocess_save_to_queue,
            args=(
                trainer.preprocess_patient,
                q,
                list_of_lists[i::num_processes],
                output_files[i::num_processes],
                segs_from_prev_stage[i::num_processes],
                classes,
                trainer.plans['transpose_forward']
            )
        )
        pr.start() ## <------------ The error is generated here!!!!!!!!!!!!!
        processes.append(pr)

    try:
        end_ctr = 0
        while end_ctr != num_processes:
            item = q.get()
            if item == "end":
                end_ctr += 1
                continue
            else:
                yield item

    finally:
        for p in processes:
            if p.is_alive():
                p.terminate()
            p.join()

        q.close()


def predict_cases_fastest(...):
    ...
    pool = Pool(num_threads_nifti_save)
	...
    preprocessing = preprocess_multithreaded(
		trainer,
		list_of_lists,
		cleaned_output_files,
		num_threads_preprocessing,
		segs_from_prev_stage
	)
	...
	pool.starmap_async(...)
	...
	pool.close()
    pool.join()


def predict_from_folder(...):
	...
    return predict_cases_fastest(...)


if __name__ == "__main__":
    ...

```

**Comment by rokm** [comment, marked as answer]
This looks like a serialization/deserialization issue...

Specifically, the long chain of calls to functions in `pickle` and `dill` modules ends up tripping [this assert](https://github.com/pyinstaller/pyinstaller/blob/d6ff0802dd6f2fbeccb4a6c47e1cf24c61c31205/PyInstaller/loader/pyimod01_archive.py#L76) in our code. This seems to indicate that something in your code (or the 3rd party packages you are using) is over-zealously pickling *everything*, including PyInstaller's `FrozenImporter`, and consequently its instance of `ZlibArchiveReader` and its `ArchiveFile`. Which is not the best idea in general, and leads to problems with the custom thread-local storage implementation that `ArchiveFile` uses.

I have a [WIP branch](https://github.com/rokm/pyinstaller/tree/cleanup-archive-writer-reader) that removes the `ArchiveFile` and associated thread-local storage as part of a wider code clean-up. Can you try using it to see if it improves the situation, or at least changes the error?

```
pip uninstall pyinstaller
pip install git+https://github.com/rokm/pyinstaller.git@cleanup-archive-writer-reader
```

**Reply by CarloPoletto** [reply]
First of all, thank you for your reply. After launching your instructions, I verified the correct installation of the branch by executing the `pip freeze --local` instruction and obtaining the following:

```
altgraph==0.17.3
batchgenerators==0.21    
certifi==2022.12.7       
charset-normalizer==3.0.1
colorama==0.4.6
contourpy==1.0.7
cx-Freeze==6.14.2        
cx-Logging==3.1.0        
cycler==0.11.0
dicom2nifti==2.4.7
dill==0.3.6
fonttools==4.38.0
fury==0.8.0
future==0.18.3
idna==3.4
imageio==2.25.0
joblib==1.2.0
kiwisolver==1.4.4
lief==0.12.3
linecache2==1.0.0
matplotlib==3.6.3
MedPy==0.4.0
multiprocess==0.70.14
networkx==3.0
nibabel==5.0.0
nnunet-customized==1.1
numpy==1.24.1
p-tqdm==1.4.0
packaging==23.0
pandas==1.5.3
pathos==0.3.0
pefile==2022.5.30
Pillow==9.4.0
pox==0.3.2
ppft==1.7.6.6
psutil==5.9.4
pydicom==2.3.1
pyinstaller @ git+https://github.com/rokm/pyinstaller.git@cb4e61c87daf3d93913a392e4f81ab4a47b39bf0
pyinstaller-hooks-contrib==2022.15
pyparsing==3.0.9
python-dateutil==2.8.2
python-gdcm==3.0.20
pytz==2022.7.1
PyWavelets==1.4.1
pywin32-ctypes==0.2.0
requests==2.28.2
scikit-image==0.19.3
scikit-learn==1.2.1
scipy==1.10.0
SimpleITK==2.2.1
six==1.16.0
sklearn==0.0.post1
threadpoolctl==3.1.0
tifffile==2023.1.23.1
torch==1.13.1+cu117
torchaudio==0.13.1+cu117
torchvision==0.14.1+cu117
tqdm==4.64.1
traceback2==1.4.0
typing-extensions==4.4.0
unittest2==1.1.0
urllib3==1.26.14
vtk==9.2.5
xvfbwrapper==0.2.9
```
After that, I recreated the .exe package as follows:
```
pyinstaller -y --clean --additional-hooks-dir=hooks --specpath %DIR_DIST% --distpath %DIR_DIST%\dist --workpath %DIR_DIST%\build -n MyProject main.py
```
Running the executable the program always seems to stop at the same point but this time the error is different:
```
Traceback (most recent call last):
  File "main.py", line 341, in <module>
  File "D:\MyProject\venv\Lib\site-packages\PyInstaller\hooks\rthooks\pyi_rth_multiprocessing.py", line 49, in _freeze_support
    spawn.spawn_main(**kwds)
  File "multiprocessing\spawn.py", line 116, in spawn_main
  File "multiprocessing\spawn.py", line 126, in _main
ModuleNotFoundError: No module named '__builtin__'
[25768] Failed to execute script 'main' due to unhandled exception!
```
Line 341 of `main.py` is as follows:
```
from multiprocessing import freeze_support

...

if __name__ == "__main__":
    freeze_support() ## <------------ Line 341
    ...
```

**Reply by rokm** [reply]
>```
>Traceback (most recent call last):
>  File "main.py", line 341, in <module>
>  File "D:\MyProject\venv\Lib\site-packages\PyInstaller\hooks\rthooks\pyi_rth_multiprocessing.py", line 49, in _freeze_support
>    spawn.spawn_main(**kwds)
>  File "multiprocessing\spawn.py", line 116, in spawn_main
>  File "multiprocessing\spawn.py", line 126, in _main
>ModuleNotFoundError: No module named '__builtin__'
>[25768] Failed to execute script 'main' due to unhandled exception!
>```

Hmm, `__builtin__` is a python2-era name for the `__builtins__` module, and is not available in python3. So it is rather confusing that it should come up at all... And the referenced `multiprocessing.spawn` line is dealing with [data unpickling](https://github.com/python/cpython/blob/db924a46d754bac6b8debffc3d9f3e9b67cbb884/Lib/multiprocessing/spawn.py#L126), so I would guess that the reference to `__builtin__` somehow ends up in pickled data (presumably coming from parent process to work process, where it causes the error).

That said, I checked `nnunet` and it does not seem to depend on `dill` - do you use it yourself to serialize/deserialize data to queue? Can you share that part of the code? Or, even better, it is possible to isolate it into a stand-alone minimal example (without `nnunet`) that still reproduces the problem? That way, I could try running it locally and see what's really going on.

**Reply by rokm** [reply]
> That said, I checked `nnunet` and it does not seem to depend on `dill` - do you use it yourself to serialize/deserialize data to queue? Can you share that part of the code? Or, even better, it is possible to isolate it into a stand-alone minimal example (without `nnunet`) that still reproduces the problem? That way, I could try running it locally and see what's really going on.

Actually, as the error is originating in the `spawn._main` that's called in the subprocess, the issue is likely not serialization/deserialization to the queue, but rather of arguments that are passed to the spawned subprocess (`[trainer.preprocess_patient, q, list_of_lists[i::num_processes], output_files[i::num_processes], segs_from_prev_stage[i::num_processes], classes, trainer.plans['transpose_forward']]`). Still, it would help if you could reduce the code into minimal self-contained example that I can work with (and in the process of reducing the example, you might also discover which argument is causing the problems with (de)serialization in the frozen application).

**Reply by CarloPoletto** [reply]
Thanks again for your support. I'm working on producing a minimal project by isolating the only part of the code that creates problems. I hope to be able to create a GitHub project by tomorrow.

**Reply by CarloPoletto** [reply]
> Actually, as the error is originating in the `spawn._main` that's called in the subprocess, the issue is likely not serialization/deserialization to the queue, but rather of arguments that are passed to the spawned subprocess (`[trainer.preprocess_patient, q, list_of_lists[i::num_processes], output_files[i::num_processes], segs_from_prev_stage[i::num_processes], classes, trainer.plans['transpose_forward']]`). Still, it would help if you could reduce the code into minimal self-contained example that I can work with (and in the process of reducing the example, you might also discover which argument is causing the problems with (de)serialization in the frozen application).

I have created a public project with which it is possible to reproduce the reported problem: https://gitlab.com/carlopoletto/nnunet_pyinstaller_problem

In the `./scripts` folder there are some scripts to install everything and run the tests:

 - `./scripts/install`: dependency installation
 - `./scripts/dist`: creating the executable with pyinstaller
 - `./scripts/run_py`: running the python script (NB: this script automatically delete the `./temp` folder and recreate it by copying the contents of `./data`)
 - `./scripts/run_exe`: running the executable created with `./scripts/dist` (NB: this script automatically delete the `./temp` folder and recreate it by copying the contents of `./data`)

The problem appears to be internal to the `nnunet` library used as a dependency. I don't know if this problem can be solved by properly configuring `pyinstaller`.

**Reply by rokm** [reply]
Thanks for setting up the full reproducer!

So first thing to note is that you are not using [nnunet](https://pypi.org/project/nnunet/) bur rather [nnunet-customized](https://pypi.org/project/nnunet-customized/). On Windows, the `nunnet.inference.predict` module of this version uses `pathos` instead of standard `multiprocessing`. I.e., [the module contains this block](https://github.com/wasserth/nnUNet_cust/blob/c64bfab04a7b2fcf93cd648f7b41b13546157a09/nnunet/inference/predict.py#L24-L30):
```python
if 'win' in sys.platform:
    #fix for windows platform
    import pathos
    Process = pathos.helpers.mp.Process
    Queue = pathos.helpers.mp.Queue
else:
    from multiprocessing import Process, Queue
```

`pathos` uses [`multiprocess`](https://pypi.org/project/multiprocess/) instead of `multiprocessing`, if available (see [here](https://github.com/uqfoundation/pathos/blob/0822da123923d19c990c288a8c6938c1b2ccb98c/pathos/helpers/__init__.py#L13-L24)). And `multiprocess` is "a fork of `multiprocessing` that extends `multiprocessing` to provide enhanced serialization, using `dill`".

So that clears up how `dill` ended up in your original error traceback; it is implicitly used for serialization because `multiprocess` primitives are used in `nunnet.inference.predict.preprocess_multithreaded`.

It also explains the second error:  the parent process is using `multiprocess` + `dill` for serialization of arguments, but the child process is using `multiprocessing` for deserialization. This latter part is not so obvious, but when you call `multiprocessing.freeze_support` in the frozen application, you are actually calling the override that was set up by [PyInstaller's `multiprocessing` run-time hook](https://github.com/pyinstaller/pyinstaller/blob/develop/PyInstaller/hooks/rthooks/pyi_rth_multiprocessing.py); this [override](https://github.com/pyinstaller/pyinstaller/blob/d6ff0802dd6f2fbeccb4a6c47e1cf24c61c31205/PyInstaller/hooks/rthooks/pyi_rth_multiprocessing.py#L23-L50) identifies the multiprocessing command-line arguments and diverts the program flow into the corresponding multiprocessing module (instead of running your main program code again).

So to make this work with `multiprocess`, you would need to create a `multiprocess` run-time hook, which is basically the same as the `multiprocessing` one, except with module name changed everywhere, e.g.:

<details><summary>src\hooks\rthooks\pyi_rth_multiprocess.py (click on triangle symbol to expand)</summary>
<p>

```python
# Use function to avoid global namespace clash with multiprocessing hook
def _pyi_rth_multiprocess():
    import multiprocess
    import multiprocess.spawn as spawn
    # 'spawn' multiprocess needs some adjustments on osx
    import os
    import sys
    from subprocess import _args_from_interpreter_flags

    # prevent spawn from trying to read __main__ in from the main script
    multiprocess.process.ORIGINAL_DIR = None


    def _freeze_support():
        print("FREEZE SUPPORT FOR MULTIPROCESS:", sys.argv)
        # We want to catch the two processes that are spawned by the multiprocess code:
        # - the semaphore tracker, which cleans up named semaphores in the spawn multiprocess mode
        # - the fork server, which keeps track of worker processes in forkserver mode.
        # both of these processes are started by spawning a new copy of the running executable, passing it the flags from
        # _args_from_interpreter_flags and then "-c" and an import statement.
        # Look for those flags and the import statement, then exec() the code ourselves.

        if (
            len(sys.argv) >= 2 and sys.argv[-2] == '-c' and sys.argv[-1].startswith((
                'from multiprocess.semaphore_tracker import main',  # Py<3.8
                'from multiprocess.resource_tracker import main',  # Py>=3.8
                'from multiprocess.forkserver import main'
            )) and set(sys.argv[1:-2]) == set(_args_from_interpreter_flags())
        ):
            exec(sys.argv[-1])
            sys.exit()

        if spawn.is_forking(sys.argv):
            kwds = {}
            for arg in sys.argv[2:]:
                name, value = arg.split('=')
                if value == 'None':
                    kwds[name] = None
                else:
                    kwds[name] = int(value)
            spawn.spawn_main(**kwds)
            sys.exit()


    multiprocess.freeze_support = spawn.freeze_support = _freeze_support

    # Bootloader unsets _MEIPASS2 for child processes to allow running PyInstaller binaries inside pyinstaller binaries.
    # This is ok for mac or unix with fork() system call. But on Windows we need to overcome missing fork() function.

    if sys.platform.startswith('win'):
        import multiprocess.popen_spawn_win32 as forking
    else:
        import multiprocess.popen_fork as forking
        import multiprocess.popen_spawn_posix as spawning


    # Mix-in to re-set _MEIPASS2 from sys._MEIPASS.
    class FrozenSupportMixIn:
        def __init__(self, *args, **kw):
            if hasattr(sys, 'frozen'):
                # We have to set original _MEIPASS2 value from sys._MEIPASS to get --onefile mode working.
                os.putenv('_MEIPASS2', sys._MEIPASS)  # @UndefinedVariable
            try:
                super().__init__(*args, **kw)
            finally:
                if hasattr(sys, 'frozen'):
                    # On some platforms (e.g. AIX) 'os.unsetenv()' is not available. In those cases we cannot delete the
                    # variable but only set it to the empty string. The bootloader can handle this case.
                    if hasattr(os, 'unsetenv'):
                        os.unsetenv('_MEIPASS2')
                    else:
                        os.putenv('_MEIPASS2', '')


    # Patch forking.Popen to re-set _MEIPASS2 from sys._MEIPASS.
    class _Popen(FrozenSupportMixIn, forking.Popen):
        pass


    forking.Popen = _Popen

    if not sys.platform.startswith('win'):
        # Patch spawning.Popen to re-set _MEIPASS2 from sys._MEIPASS.
        class _Spawning_Popen(FrozenSupportMixIn, spawning.Popen):
            pass

        spawning.Popen = _Spawning_Popen


_pyi_rth_multiprocess()
del _pyi_rth_multiprocess
```

</p></details>

The runtime hooks are ran as scripts, so their symbols are stored in global namespace and can be overwritten by different hooks. That's why compared to original `multiprocessing` runtime hook, this one puts everything in a function, to avoid symbol clashes with the original runtime hook. (Otherwise the `spawn` symbol (= `multiprocess.spawn`) gets overwritten by `multiprocessing.spawn` one when `multiprocessing` runtime hook is ran after this `multiprocess` one, which effectively restores original deserialization problem).

In addition to above runtime hook being added to your custom hooks directory, you also need the .dat file that maps module name to runtime hook:

<details><summary>src\hooks\rthooks.dat (click on triangle symbol to expand)</summary>
<p>

```
{
    'multiprocess': ['pyi_rth_multiprocess.py'],
}
```

</p></details>

Finally, you need to ensure that you call `multiprocess.freeze_support` instead of `multiprocessing.freeze_support`, by modifying `main.py`. I think the safest way to do that is to import from `pathos`:

<details><summary>modified src\main.py (click on triangle symbol to expand)</summary>
<p>

```python
# Deal with frozen multiprocessing support before importing nnunet
if __name__ == "__main__":
    # On Windows, nnunet-custom uses pathos (if available), which in turn uses multiprocess fork of multiprocessing
    import sys
    if 'win' in sys.platform:
        try:
            from pathos.helpers import freeze_support
        except Exception as e:
            from multiprocessing import freeze_support
    else:
        from multiprocessing import freeze_support
    freeze_support()

import shutil
import argparse

from pathlib import Path
from distutils.dir_util import copy_tree
from nnunet.inference.predict import predict_from_folder

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-data",
        dest="data",
        metavar="filepath",
        type=lambda p: Path(p).absolute(),
        required=True
    )
    
    dir_data = parser.parse_args().data

    predict_from_folder(
        model=str(dir_data.joinpath("model")),
        input_folder=str(dir_data.joinpath("input")),
        output_folder=str(dir_data.joinpath("input")),
        folds=[0],
        save_npz=False,
        num_threads_preprocessing=6,
        num_threads_nifti_save=2,
        lowres_segmentations=None,
        part_id=0,
        num_parts=1,
        tta=False,
        overwrite_existing=False,
        mode="fastest",
        overwrite_all_in_gpu=None,  # type: ignore
        mixed_precision=True,
        step_size=0.5,
        checkpoint_name="model_final_checkpoint",
    )
```

</p></details>

Ideally, you would call both `multiprocessing.freeze_support` and `multiprocess.freeze_support`; however both modules use identical command-line arguments for their subprocesses, so there is no way to identify which module they are coming from. That's why in above example, we use only `multiprocess.freeze_support` (via `pathos`); this implicitly assumes that if parent process uses `multiprocessing` (which other `nnunet` modules seem to do...) for spawning the process and serialization, the `multiprocess` + `dill` deserialization can actually take care of it (i.e, is backwards-compatible with standard `multiprocessing` serialization format). Which might actually be true (while the reverse obviously is not), or may lead to different problems. But the sample program seems to work now.

**Reply by CarloPoletto** [reply]
Thank you very much. I have done several tests and using your WIP branch and your modifications described above everything seems to work very well. Will the WIP branch be included in the main branch in the future?

**Reply by rokm** [reply]
> Will the WIP branch be included in the main branch in the future?

That's the plan; it just need to be finished first (hence work-in-progress).
